### PR TITLE
Damo/load balancer

### LINKED ISSUE #3: Socket opening exhausts port connections for high workloads files.
The workload_gen.py and performRequest function continually open sockets for each user command -- with higher workloads this exhausts all open ports in the system resulting in error: 

```
ERROR:root:Bind failed. Error: [WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted
```

**Notes: **
 * Close is used on the socket after sending, but this functionality will only destroy socket after an OS routine has responded to it (~delay of 30 secs.) from close.
 * Found bug while making a load_balancer.py server with the same socket usage functionality. 

see below image for an example of open ports 

image

### PR DESCRIPTION
**Pull request by DamonAknuh**
load balancer server

splits incoming requests into the available webservers

### CODE REVIEW

File: load_balancer/src/load_balancer.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+# Table of open web servers to forward packet requests too. 
+#    - user IDs are key- hashed and modulos with NUM of servers
+#       this determines the server index to forward user request too
+#    - Audit server is always last array index. 
+forwardServers = []
+
+PACKET_SIZE     = 8192 
+
+class webServer():
```
Comment (skosman) [inline load_balancer/src/load_balancer.py]: Is this supposed to be loadBalancer instead of webServer?

**Review by skosman — APPROVED** [review]

File: load_balancer/src/load_balancer.py (hunk 1 above)
Comment (DamonAknuh) [reply load_balancer/src/load_balancer.py]: Changed in next commit

File: load_balancer/src/load_balancer.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        while self.serverRunning:
+            try: 
+                packet = self.packetQ.get(False)
+
+                if "user" in packet.keys():
+                    user = packet["user"]
+                else:
+                    user = "default"
+                
+                serverHash = hash(user) % NUM_FORWARD_SERVERS
```
Comment (DiannaBanana) [inline load_balancer/src/load_balancer.py:182]: this is the constant hashing of the balancer right?

**Review by DiannaBanana — APPROVED** [review]

File: load_balancer/src/load_balancer.py (hunk 2 above)
Comment (DamonAknuh) [reply load_balancer/src/load_balancer.py:182]: Yep! 
It's constant for the application runtime. If you restart the LB the same user might be directed to a different server.

Comment (DiannaBanana) [reply load_balancer/src/load_balancer.py:182]: Ahh nice work! We can test it more exhaustively when we have our apps more clearly separated into services

File: backend/src/user_commands.py [hunk 3]
```diff
@@ -304,20 +296,13 @@ def CMD_CommitSell(cmdDict, threadContext):
...
     cmdDict['timestamp'] = str(int(time.time()*1000))
-
-    if not threadContext["sellAmtQ"].empty():
-        threadContext["sellAmtQ"].put(cmdDict['amount'])
-        log.logEvents['userCommand'](cmdDict)
-    else: 
-        cmdDict['amount'] = "0.00"
-        cmdDict['errorMessage'] = "Invalid cmd. No recent pending buys" 
-
+    log.logEvents['userCommand'](cmdDict)
```
Comment (DamonAknuh) [inline backend/src/user_commands.py:299]: Dianna. It seems that we don't have a sellAmtQ, and buyAmtQ anymore.  Do we still need this line? 
`log.logEvents['userCommand'](cmdDict)`

I moved those two queues into the user thread contexts. Now it seems with Sams changes these queues aren't needed?

Comment (DiannaBanana) [reply backend/src/user_commands.py:299]: I think we still need to log those commands tho

Comment (DamonAknuh) [reply backend/src/user_commands.py:299]: which commands? the else branch?

```
    else: 	
        cmdDict['amount'] = "0.00"	
        cmdDict['errorMessage'] = "Invalid cmd. No recent pending buys"
```

Comment (DiannaBanana) [reply backend/src/user_commands.py:299]: I'm still figuring my way around mongo and sams changes, but I believe the triggers are handled in the db so yea the queues might not be needed , you're right

Comment (DiannaBanana) [reply backend/src/user_commands.py:299]: I think the log is outside of the if else, so it logs it regardless. But the logging line can probably be moved to a better spot

Comment (DamonAknuh) [reply backend/src/user_commands.py:299]: Sweet, thanks!  I'll keep them removed and hope nothing breaks :)

Comment (DiannaBanana) [reply backend/src/user_commands.py:299]: :) fingers crossed!

**Review by DanielZenk — APPROVED** [review]

**Pull request merged**